File: python/test_heap.py
import pytest
import heapq
import heap as h

@pytest.fixture
def ary():
    return [2, 3, 1, 4, 5]

def test_sort_heap(ary):
    assert h.sort_heap(ary) == [1, 2, 3, 4, 5]

def test_is_heap(ary):
    assert not h.is_heap(ary)
    heapq.heapify(ary)
    assert h.is_heap(ary)
    assert h.is_heap([2, 3, 4, 4, 4])

# is_heap_until return the index of first element
# which disobey the requirement of heap.
def test_is_heap_until(ary):
    assert h.is_heap_until(ary) == 2
    heapq.heapify(ary)
    assert h.is_heap_until(ary) == len(ary)

def test_pop_heap(ary):
    heapq.heapify(ary)
    min_value = h.pop_heap(ary)
    assert min_value == 1

def test_push_heap(ary):
    heapq.heapify(ary)
    h.push_heap(ary, 3)
    assert h.is_heap(ary)

def test_make_head(ary):
    h.make_heap(ary)
    assert h.is_heap(ary)

File: ruby/test_heap.rb
require 'minitest/autorun'
require_relative 'heap'

class TestHeap < MiniTest::Test
  include Heap

  def setup
    @a = [2, 3, 1, 4, 5]
    @h = [1, 2, 3, 5, 4]
  end

  def test_is_heap
    assert_equal false, is_heap(@a)
    assert_equal true, is_heap(@h)
  end

  def test_is_heap_until
    assert_equal 2, is_heap_until(@a)
    assert_equal @h.size, is_heap_until(@h)
  end

  def test_sort_heap
    assert_equal [1, 2, 3, 4, 5], sort_heap(@a)
  end

  def test_pop_heap
    make_heap @a
    assert_equal 1, pop_heap(@a)
  end

  def test_push_heap
    push_heap(@h, 3)
    assert_equal true, is_heap(@h)
  end

  def test_make_heap
    make_heap @a
    assert_equal true, is_heap(@a)

    make_heap @h
    assert_equal true, is_heap(@h)
  end
end
